### PR TITLE
DMA Locking with Thread Priority Inheritance

### PR DESCRIPTION
**Pull request by andyp1per**
This PR converts the lock in Shared_DMA to a mutex allowing thread priority inheritance on threads taking out DMA locks. 

To do this requires locking and unlocking from the same thread and also requires that locks are unlocked in reversed order of locking. 

In order to satisfy the thread requirement, UART TX and rcout have been moved to their own threads and an event-based model used to request that the thread perform an action.

In order to satisfy the ordering requirement, DMA unlocking has meed moved from interrupt handlers into the appropriate thread and the interrupt handler now requests an unlock. 

With a separate thread it is now possible for both rcout and UART tx to take out the DMA lock blocking.

Flown (briefly) on a Kakute F7 mini with dshot 600. Also tested on a Durandal with CRSF (so DMA to the IOMCU and to the CRSF RX). Flown (hover) on a Pixracer with OneShot125

The larger number of threads consume stack so I have trimmed some of the threads in question and reduced ```PORT_INT_REQUIRED_STACK``` to 128

The reason for doing this is to reduced shared DMA contention so that bi-directional dshot can operate more effectively.

To aid debugging I added @SYS/uarts.txt:

```
UARTV1
OTG1     TX =  240588 RX =   13521 TXBD= 21077 RXBD=  1184
UART4    TX*=    1022 RX*=       0 TXBD=    89 RXBD=     0
UART2    TX*=  215147 RX*=       0 TXBD= 18848 RXBD=     0
UART3    TX*=  202850 RX*=       0 TXBD= 17770 RXBD=     0
UART8    TX*=      81 RX =      24 TXBD=     7 RXBD=     2
UART1    TX =      14 RX =  313950 TXBD=     1 RXBD= 27504
UART7    TX =       0 RX =       0 TXBD=     0 RXBD=     0
```

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/hwdef/f303-MatekGPS/hwdef-bl.dat [hunk 1]
```diff
@@ -61,6 +61,7 @@ define PORT_INT_REQUIRED_STACK 64
 
 # avoid timer and RCIN threads to save memory
 define HAL_NO_TIMER_THREAD
+define HAL_NO_RCOUT_THREAD
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/f303-MatekGPS/hwdef-bl.dat:64]: move all 3 of these defines for all bootloaders to chibios_hwdef.py

**Review by tridge — CHANGES_REQUESTED** [review]
main thing needed is lots of testing!

**Comment by andyp1per** [comment]
Flew this hard on a 4" Pixracer using all its UARTs. Flight and performance was good - so I think the code is working correctly, however there is definitely a problem with thread priority for logging. I got a logging failed once and you can see from the logs that the logging is not able to keep up even though this is a very fast SD card. Also flying a similar change on a KakuteF7mini with flash logging I saw a similar issue - I could start chip erase but it would not complete, indicating that the logging thread is not getting enough cycles. I'm not sure what the right answer is here, I suspect the unbuffered UART priority is contributing to this.

**Comment by tridge** [comment]
on memory, worth seeing if reducing SERIAL_BUFFERS_SIZE is OK now we have separate thread per uart, plus reduce the min buffer sizes in UARTDriver.cpp, and reduce the special case multipler for USB

**Comment by andyp1per** [comment]
Rebased

File: libraries/AP_HAL_ChibiOS/hwdef/f303-MatekGPS/hwdef-bl.dat (hunk 1 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/f303-MatekGPS/hwdef-bl.dat]: Fixed

**Comment by andyp1per** [comment]
Flew this without issue on a KakuteF7Mini. (6 packs) I had to adjust the DMA a little because I was using bi-dir and the dataflash was contending with the IMU, but the good news is I was able to share the dataflash with TIM1_UP (dshot output) without problems. The new locking strategy means that this is much safer to do now.

[6 01-01-1970 00-00-01.zip](https://github.com/ArduPilot/ardupilot/files/5790872/6.01-01-1970.00-00-01.zip)

**Comment by tridge** [comment]
I flew this today on a plane with no issues

**Comment by andyp1per** [comment]
There is still an issue with H743 boards that I am investigating.

**Comment by andyp1per** [comment]
Rebased

**Comment by amilcarlucas** [comment]
What is the status of this? it will probably be a big help for https://github.com/ArduPilot/ardupilot/pull/14678 
Can we discuss this on the next dev call

**Comment by andyp1per** [comment]
@amilcarlucas it's ready to go - it just needs the review green light from @tridge

**Comment by amilcarlucas** [comment]
Opps, there is a merge conflict now.

**Comment by andyp1per** [comment]
Easily fixed.

**Comment by amilcarlucas** [comment]
Thanks!

**Comment by andyp1per** [comment]
Built iofirmware and tested that that still works

**Comment by andyp1per** [comment]
Flew 6 batteries on this pretty hard without issues. KakuteF7Mini - 3 connected uarts and 5 connected uarts, dshot600

File: libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp [hunk 2]
```diff
@@ -480,7 +481,11 @@ void RCOutput::dma_up_irq_callback(void *p, uint32_t flags)
         TOGGLE_PIN_DEBUG(55);
     } else {
         // non-bidir case
-        chVTSetI(&group->dma_timeout, chTimeUS2I(group->dshot_pulse_time_us + 40), dma_unlock, p);
+        // this prevents us ever having two dshot pulses too close together
+        // LED sends might be quite long so adjust the timeout appropriately
+        const uint32_t period_us = AP_HAL::micros() - group->last_dmar_send_us;
+        const uint32_t remain_us = period_us < group->dshot_pulse_time_us ? 40 + group->dshot_pulse_time_us - period_us : 40;
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp:487]: the wrap handling with this 40+ isn't clear. Better to calculate remain_us then add 40 in the chVTSetI() call

File: libraries/AP_HAL_ChibiOS/UARTDriver.cpp [hunk 3]
```diff
@@ -83,51 +97,113 @@ _baudrate(57600)
...
-  initialise UART thread
+  initialise UART RX thread
  */
-void UARTDriver::thread_init(void)
+void UARTDriver::thread_rx_init(void)
 {
-    if (uart_thread_ctx) {
-        // already initialised
-        return;
+#if CH_CFG_USE_HEAP == TRUE
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/UARTDriver.cpp:167]: better to not have this #if - we'd rather not compile than silently fail to work. We need heap for everything other than bootloader, and the bootloader sets HAL_NO_UARTDRIVER

File: libraries/AP_HAL_ChibiOS/UARTDriver.cpp [hunk 4]
```diff
@@ -83,51 +97,113 @@ _baudrate(57600)
...
     }
+#endif
+}
+
+/*
+  initialise UART TX_thread
+ */
+void UARTDriver::thread_init(void)
+{
 #if CH_CFG_USE_HEAP == TRUE
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/UARTDriver.cpp:186]: same here, don't check CH_CFG_USE_HEAP

**Review by tridge — CHANGES_REQUESTED** [review]
still more to look at, but this is a start. Sorry it took so long

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 5]
```diff
@@ -110,6 +116,75 @@ void RCOutput::init()
...
+            // above 1000 Hz. This makes the protocol more reliable on
+            // long cables, and also keeps some ESCs happy that don't
+            // like low rates
+            dshot_send_groups();
+            // now unlock everything
+            dshot_collect_dma_locks(last_thread_run_us);
+        }
+
+        // process any pending RC output requests
+        timer_tick(last_thread_run_us);
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp:153]: Yes its triggered by EVT_DSHOT_SEND which maybe we should rename

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 6]
```diff
@@ -61,9 +64,12 @@ const uint8_t RCOutput::NUM_GROUPS = ARRAY_SIZE(RCOutput::pwm_group_list);
 void RCOutput::init()
 {
     uint8_t pwm_count = AP_BoardConfig::get_pwm_count();
-    for (auto &group : pwm_group_list) {
+    for (uint8_t i = 0; i < NUM_GROUPS; i++) {
+        pwm_group &group = pwm_group_list[i];
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp]: Fixed

File: libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp (hunk 2 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp]: Fixed

**Comment by andyp1per** [comment]
@tridge I have increased PORT_INT_REQUIRED_STACK to 128. I don't think we can set SERIAL_BUFFERS_SIZE smaller as the RX thread still handles all UARTS and there doesn't seem to be a way of setting the send buffer independently from the receive buffer.

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 7]
```diff
@@ -110,6 +116,75 @@ void RCOutput::init()
...
+
+        // process any pending RC output requests
+        timer_tick(last_thread_run_us);
+    }
+}
+
+// release locks on the groups that are pending in reverse order
+void RCOutput::dshot_collect_dma_locks(uint32_t last_run_us)
+{
+    for (int8_t i = NUM_GROUPS - 1; i >= 0; i--) {
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp:160]: After talking with @peterbarker I have reverted this particular change as an iterator is not appropriate in this case

**Comment by andyp1per** [comment]
On the current build, I have the following memory free

|Board|master|This PR|
|---|---|---
|F405|10616|7208|
|F745(5UARTS)|121192|116880|
|F745(3UARTS)|129232|124928|

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 8]
```diff
@@ -110,6 +116,80 @@ void RCOutput::init()
...
+        pwm_group &group = pwm_group_list[i];
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:161]:
```suggestion
    for (uint8_t i = 0; i < NUM_GROUPS; i++) {
        const uint8_t index = NUM__GROUPS - i;
        pwm_group &group = pwm_group_list[index];
```

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 9]
```diff
@@ -110,6 +116,80 @@ void RCOutput::init()
...
+
+        // process any pending RC output requests
+        timer_tick(last_thread_run_us);
+    }
+}
+
+// release locks on the groups that are pending in reverse order
+void RCOutput::dshot_collect_dma_locks(uint32_t last_run_us)
+{
+    for (int8_t i = NUM_GROUPS - 1; i >= 0; i--) {
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:160]:
```suggestion
    if (NUM_GROUPS == 0) {
        return;
    }
    for (int8_t i = NUM_GROUPS - 1; i >= 0; i--) {
```

**Comment by andyp1per** [comment]
@tridge I have trimmed stack from uart_rx, rcout and storage, saving 1.5k. I think this is safe as no-one is randomly using these threads and each still has at least 450bytes free.

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
>  I don't think we can set SERIAL_BUFFERS_SIZE smaller as the RX thread still handles all UARTS and there doesn't seem to be a way of setting the send buffer independently from the receive buffer.

I did end up lowering it after doing some testing. It seems to perform fine with a lower size

**Comment by tridge** [comment]
i'm happy for this to be merged when it passes CI.

**Comment by andyp1per** [comment]
Fantastic! Thanks for all your help @tridge

**Pull request merged**